tests/test_web_poet_rules.py

```diff
@@ -1467,28 +1471,148 @@ def test_page_object_with_item_dependency_deadlock_2_a(caplog) -> None:
+
+    def __call__(self, to_provide: Set[Callable], request: scrapy.Request):
+        return [Mobius(name="mobius from MobiusProvider")]
+
+
+@inlineCallbacks
+def test_page_object_returning_item_which_is_also_a_dep() -> None:
+    """Tests that the item from a provider has been modified correctly
+    by the corresponding PO.
+    """
+
+    settings = {
+        "SCRAPY_POET_PROVIDERS": {**DEFAULT_PROVIDERS, **{MobiusProvider: 1000}}
+    }
+    # item from 'to_return'
+    item, deps = yield crawl_item_and_deps(Mobius, override_settings=settings)
+    assert item == Mobius(name="(modified) mobius from MobiusProvider")
+    assert_deps(deps, {"item": Mobius})
+
+    # calling the actual page objects should still work
+    item, deps = yield crawl_item_and_deps(MobiusPage, override_settings=settings)
+    assert item == Mobius(name="(modified) mobius from MobiusProvider")
+    assert_deps(deps, {"page": MobiusPage})
+
+
+@inlineCallbacks
+def test_page_object_returning_item_which_is_also_a_dep_but_no_provider_item(
+    caplog,
+) -> None:
+    """Same as ``test_page_object_returning_item_which_is_also_a_dep()``
+    but there's no provider for the original item
+    """
+    yield crawl_item_and_deps(Mobius)
+    assert "ProviderDependencyDeadlockError" in caplog.text
+
+
+@inlineCallbacks
+def test_page_object_returning_item_which_is_also_a_dep_but_no_provider_po(
+    caplog,
+) -> None:
+    """Same with ``test_page_object_returning_item_which_is_also_a_dep_but_no_provider_item()``
+    but tests the PO instead of the item.
+    """
+    yield crawl_item_and_deps(MobiusPage)
+    assert "ProviderDependencyDeadlockError" in caplog.text
+
+
+@attrs.define
+class Kangaroo:
+    name: str
+
+
+@handle_urls(URL)
+@attrs.define
+class KangarooPage(WebPage[Kangaroo]):
+    kangaroo_item: Kangaroo
+
+    @field
+    def name(self) -> str:
+        return f"(modified) {self.kangaroo_item.name}"
+
+
+@handle_urls(URL, priority=600)
+@attrs.define
+class JoeyPage(KangarooPage):
+    @field
+    def name(self) -> str:
+        return f"(modified by Joey) {self.kangaroo_item.name}"
+
+
+class KangarooProvider(PageObjectInputProvider):
+    provided_classes = {Kangaroo}
+
+    def __call__(self, to_provide: Set[Callable], request: scrapy.Request):
+        return [Kangaroo(name="data from KangarooProvider")]
+
+
+@inlineCallbacks
+def test_page_object_returning_item_which_is_also_a_dep_2() -> None:
+    """Same with ``test_page_object_returning_item_which_is_also_a_dep()`` but
+    there are now 2 POs with different priorities that returns the item.
+
+    The PO with the higher priority should be the one returning the item.
+    """
+
+    settings = {
+        "SCRAPY_POET_PROVIDERS": {**DEFAULT_PROVIDERS, **{KangarooProvider: 1000}}
+    }
+    # item from 'to_return'
+    item, deps = yield crawl_item_and_deps(Kangaroo, override_settings=settings)
+    assert item == Kangaroo(name="(modified by Joey) data from KangarooProvider")
```

Comment:
Notice that only `JoeyPage` had touched the data and `KangarooPage` didn't because the earlier PO had higher priority.